presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1202,9 +1205,30 @@ public PlanWithProperties visitUnion(UnionNode node, PreferredProperties parentP
                 // children partitioning and don't GATHER partitioned inputs
                 // TODO: add FIXED_ARBITRARY_DISTRIBUTION support on non empty singleNodeChildren
                 if (!parentPartitioningPreference.isPresent() || parentPartitioningPreference.get().isDistributed()) {
-                    return arbitraryDistributeUnion(node, distributedChildren, distributedOutputLayouts);
+                    // TODO: can we insert LOCAL exchange for one child SOURCE distributed and another HASH distributed?
+                    if (getNumberOfTableScans(distributedChildren) == 0 && isSameOrSystemCompatiblePartitions(extractRemoteExchangePartitioningHandles(distributedChildren))) {
+                        // No source distributed child, we can use insert LOCAL exchange
+                        // TODO: if all children have the same partitioning, pass this partitioning to the parent
+                        // instead of "arbitraryPartition".
+                        return new PlanWithProperties(node.replaceChildren(distributedChildren));
+                    }
+                    else if (preferDistributedUnion) {
+                        // Presto currently can not execute stage that has multiple table scans, so in that case
+                        // we have to insert REMOTE exchange with FIXED_ARBITRARY_DISTRIBUTION instead of local exchange
+                        return new PlanWithProperties(
+                                new ExchangeNode(
+                                        idAllocator.getNextId(),
+                                        REPARTITION,
+                                        REMOTE_STREAMING,
+                                        new PartitioningScheme(Partitioning.create(FIXED_ARBITRARY_DISTRIBUTION, ImmutableList.of()), node.getOutputVariables()),
+                                        distributedChildren,
+                                        distributedOutputLayouts,
+                                        false,
+                                        Optional.empty()));
+                    }
                 }
-
+                // TODO: We should support multiple table scans in a single fragment for efficient union implementation
+                // TODO: Multiple table scans are already partially supported (for partitioned (bucketed) tables)
```

Comment:
Before, when will this branch being triggered? -- one case I can think is to gather all the table write information ?

Reply:
Yeah, table write information is one of them. Another good example is `sort`.